mindmeld/auto_annotator.py

```diff
@@ -835,5 +839,300 @@ def valid_entity_check(self, entity):
+            language (str): Language code for the language of the given sentence
+        Returns: entities (list): List of duckling candidates.
+        """
+        candidates = self.en_annotator.duckling.get_candidates_for_text(
+            sentence, language=language
+        )
+        final_spans = MultiLingualAnnotator._get_final_spans(candidates, sentence)
+        selected_candidates = []
+        for span in final_spans:
+            for candidate in candidates:
+                if span == (candidate["start"], candidate["end"]):
+                    if not (
+                        candidate["dim"] == "amount-of-money"
+                        and candidate["value"]["unit"] == "unknown"
+                    ):
+                        selected_candidates.append(candidate)
+                        break
+        return selected_candidates
+
+    @staticmethod
+    def convert_duckling_candidates(duckling_candidates):
+        """Converts duckling candidates into entity dictionaries.
+        Args:
+            duckling_candidates (list): List of Duckling candidates
+        Returns:
+            entities (list): List of converted entities.
+        """
+        return [
+            {
+                "body": entity["body"],
+                "start": entity["start"],
+                "end": entity["end"],
+                "dim": entity["entity_type"],
+                "value": entity["value"],
+            }
+            for entity in duckling_candidates
+        ]
+
+    @property
+    def supported_entity_types(self):  # pylint: disable=W0236
+        """
+        Returns:
+            supported_entity_types (list): List of supported entity types.
+        """
+        return SPACY_ANNOTATOR_SUPPORTED_ENTITIES
+
+    @staticmethod
+    def _has_overlap(span_one, span_two):
+        """Determines whether two spans overlap. Spans are two-element tuples with the format:
+        (start_index, end_index + 1).
+        """
+        return span_one[1] > span_two[0] and span_two[1] > span_one[0]
+
+    @staticmethod
+    def _is_larger(span_one, span_two):
+        """Determines whether the first spans is larger than the second. Spans are two-element
+        tuples with the format: (start_index, end_index + 1).
+        """
+        return span_one[1] - span_two[0] > span_one[1] - span_two[0]
+
+    @staticmethod
+    def _filter_mid_token_spans(spans, sentence):
+        """Removes any spans that end in-between a token.
+
+        Args:
+            spans (list): List of tuples representing candidate spans (start_index, end_index + 1).
+            sentence (str): The original sentence parsed by duckling.
+        Returns:
+            filtered_spans (list): List of spans that do not end mid-token.
+        """
+        space_indices = [i for i in range(len(sentence)) if sentence[i] == " "]
+        return [
+            span
+            for span in spans
+            if span[1] in space_indices or span[1] >= len(sentence) - 2
```

Comment:
`if end_index in space_indices or end_index >= len(sentence) - 2`

Reply:
I don't understand `end_index >= len(sentence) - 2`. Also, if a space is in-between a span, shouldnt you reject the span? ie even span index from start_index to end_index needs to be checked against space_indices.

Reply:
It's okay to have a a space in-between a span (ex: "dos horas"). Removed the filter function since it is not always valid. There are many cases of punctuation appearing right after the end of a span, here the spans should not be filtered out. (ex: "{en la mañana}?" or "{a las 2 PM|sys_time},")